CI: add release on macos-14

LeoHsiao1 opened the pull request:
Good morning!
In 2020, apple sells a new generation of mac computers, changing the CPU chips from Intel to M1, and the CPU architecture from amd64 to arm64.
In https://github.com/Exiv2/exiv2/issues/2173, some users want to use exiv2 on MacOS M1 computers. I would like to help them, although I don't have a MacOS computer.

In 2024, https://github.com/actions/runner-images/issues/9255, GitHub Actions added the macos-14 runner. It only uses M1 chips, not Intel chips.
So I modified the exiv2 workflow script to compile on macos-14. It looks like it compiled successfully, but not yet tested.

There is another problem, we need to change the filename format of the release packages, to differentiate between amd64 and arm64.
https://github.com/Exiv2/exiv2/blob/77915ad17b103793017e2b3b5ffd6c8a3ab1e916/.github/workflows/release.yml#L224-227

LeoHsiao1 committed:
CI: add release on macos-14

kmilos commented:
Thanks for the initiative! For the arch and archive names, you can check what we did for darktable recently: https://github.com/darktable-org/darktable/blob/master/.github/workflows/nightly.yml

LeoHsiao1 committed:
CI: declare the architecture of MacOS

LeoHsiao1 committed:
CI: convert ARM64 to lowercase

LeoHsiao1 commented:
Thanks for the darktable example.
In exiv2, `.github/workflows/release.yml` now generates 4 zip files:
```sh
exiv2-linux64.zip
exiv2-win.zip
exiv2-macos-12-x64.zip    # it is compatible with macos ≥12 , amd64
exiv2-macos-14-arm64.zip  # it is compatible with macos ≥14 , arm64
```

However, if you unzip `exiv2-macos-12-x64.zip` or `exiv2-macos-14-arm64.zip`, you will get a file named `exiv2-1.00.0.9-Darwin.tar.gz`.
This name is determined by cmake.
Honestly, I don't know how to configure cmake.

neheb commented on .github/workflows/release.yml:
```diff
@@ -55,7 +55,13 @@ jobs:
+      fail-fast: false
+      matrix:
+        runner:
+          - { os: macos-12, arch: X64   }
```
These were now updated.

ferdnyc committed:
CMake: Add architecture to package filename

Use `64bit` or `32bit` on Windows hosts (including MinGW/Cygwin/MSYS),
the value of `${CMAKE_HOST_SYSTEM_ARCHITECTURE}` on all others.
Always insert into package filename directly after version,
separated from preceding version and following flags with hyphens.

So, e.g.
* `exiv2-1.0.9-Linux64.tar.gz`
* `exiv2-1.0.9-Darwin.tar.gz`
* `exiv2-1.0.9-2019msvc64.zip`
become:
* `exiv2-1.0.9-x86_64-Linux.tar.gz`
* `exiv2-1.0.9-arm64-Darwin.tar.gz`
* `exiv2-1.0.9-64bit-2019msvc.zip`
respectively.

ferdnyc commented:
@LeoHsiao1 I've used one of my favorite tricks, and opened a PR against your PR.

PR LeoHsiao1/exiv2#1 will, if you merge it, incorporate my change to the packaging configs into _this_ PR. Here's my commit message explaining how the package-naming changes:

> CMake: Add architecture to package filename
> 
> Use `64bit` or `32bit` on Windows hosts (including MinGW/Cygwin/MSYS), the value of `${CMAKE_HOST_SYSTEM_ARCHITECTURE}` on all others. Always insert into package filename directly after version, separated from preceding version and following flags with hyphens.
> 
> So, e.g.
> * `exiv2-1.0.9-Linux64.tar.gz`
> * `exiv2-1.0.9-Darwin.tar.gz`
> * `exiv2-1.0.9-2019msvc64.zip`
> 
> become:
> * `exiv2-1.0.9-x86_64-Linux.tar.gz`
> * `exiv2-1.0.9-arm64-Darwin.tar.gz`
> * `exiv2-1.0.9-64bit-2019msvc.zip`
>  
> respectively.

(Committers here can then still review that change as part of this PR.)

LeoHsiao1 committed:
Merge pull request #1 from ferdnyc/macos-package-arch

CMake: Add architecture to package filename

LeoHsiao1 commented:
Thanks @ferdnyc , I merged your PR.
I run `.github/workflows/release.yml` again. Now it generate 4 files:
```sh
exiv2-linux64.zip         -> unzip ->    exiv2-1.00.0.9-x86_64-Linux.tar.gz
exiv2-macos-12-x64.zip    -> unzip ->    exiv2-1.00.0.9-x86_64-Darwin.tar.gz
exiv2-macos-14-arm64.zip  -> unzip ->    exiv2-1.00.0.9-arm64-Darwin.tar.gz
exiv2-win.zip             -> unzip ->    exiv2-1.00.0.9-64bit-2019msvc.zip
```

ferdnyc commented:
@kmilos

> However, I suggest using real arch for Windows builds as well, as WoA adoption seems to be around the corner.

That's a good point, I'm just not sure how to _do_ it for Windows. The CMake script uses the size of a void pointer in the current compiler, to determine 32- vs. 64-bit builds, and I'm pretty sure that even when `sizeof(void_p) == 4`, the `${CMAKE_HOST_SYSTEM_PROCESSOR}` will be whatever 64-bit CPU the build is actually running on.

I suppose I could just hard-code it to `i686` for 32-bit builds, since that's the only possibility, and use the real `${CMAKE_HOST_SYSTEM_PROCESSOR}` value for 64-bit builds. How do people feel about that?

ferdnyc commented on cmake/packaging.cmake:
```diff
@@ -12,13 +12,16 @@ else()
 
-set (BS "") # Bit Size
-if ( NOT APPLE )
+set (BARCH "") # Target architecture
+if ( MSVC OR MSYS OR CYGWIN OR MINGW )
   if ( CMAKE_SIZEOF_VOID_P EQUAL 8 )
```
```suggestion
set (BARCH ${CMAKE_HOST_SYSTEM_PROCESSOR}) # Target architecture
if ( CMAKE_SIZEOF_VOID_P EQUAL 4 )
  # 32-bit build, force architecture
  set (BARCH "i686")
```

ferdnyc commented on cmake/packaging.cmake:
```diff
@@ -12,13 +12,16 @@ else()
+if ( MSVC OR MSYS OR CYGWIN OR MINGW )
   if ( CMAKE_SIZEOF_VOID_P EQUAL 8 )
-    set (BS 64)
+    set (BARCH 64bit)
   else()
```
```suggestion
```

ferdnyc commented on cmake/packaging.cmake:
```diff
@@ -12,13 +12,16 @@ else()
+    set (BARCH 64bit)
   else()
-    set (BS 32)
+    set (BARCH 32bit)
   endif()
+else()
+  # Will be "arm64", "x86_64" / "amd64", etc.
+  set (BARCH ${CMAKE_HOST_SYSTEM_PROCESSOR})
```
```suggestion
```

ferdnyc reviewed (commented):
Three suggested changes that, applied together, would implement the "force i686 for 32-bit builds but use the real processor for all others" idea I mentioned in the comments.

(I had to split it up like this because of GitHub's annoying "no suggestions on ranges which include deleted lines" nonsense.)

LeoHsiao1 committed:
CI: use real arch for Windows builds

Co-authored-by: Frank Dana <ferdnyc@gmail.com>

LeoHsiao1 committed:
CI: standardize the naming format of github artifacts

LeoHsiao1 committed:
CI: fix the syntax of Github matrix.runner

LeoHsiao1 commented:
Fine, now `.github/workflows/release.yml` generate these files:
```sh
exiv2-arm64-macos-14.zip    -> unzip ->   exiv2-1.00.0.9-arm64-Darwin.tar.gz
exiv2-x64-macos-12.zip      -> unzip ->   exiv2-1.00.0.9-x86_64-Darwin.tar.gz
exiv2-x64-ubuntu-22.04.zip  -> unzip ->   exiv2-1.00.0.9-x86_64-Linux.tar.gz
exiv2-x64-windows-2022.zip  -> unzip ->   exiv2-1.00.0.9-AMD64-2019msvc.zip
```
The filename on the left side, indicates which machine exiv2 was built from.
Github Action may offer linux/arm64 and windows/arm64 in the future.

LeoHsiao1 committed:
Merge branch 'Exiv2:main' into main

LeoHsiao1 committed:
Merge branch 'Exiv2:main' into main

LeoHsiao1 commented:
This PR only changes two files, which can be merged or rebased.
